sygnal/gcmpushkin.py

```diff
@@ -464,21 +493,22 @@ def _handle_v1_response(
                 f"Unknown GCM response code {response.code}"
             )
 
-    def _get_access_token(self) -> str:
-        """Retrieve a valid access token that can be used to authorize requests.
+    async def _get_auth_header(self) -> str:
+        """Retrieve the auth header that can be used to authorize requests.
 
-        :return: Access token.
+        :return: Needed content of the `Authorization` header
         """
-        # TODO: Should we use the environment variable approach instead?
-        # export GOOGLE_APPLICATION_CREDENTIALS=/path/to/key.json
-        # credentials, project = google.auth.default(scopes=AUTH_SCOPES)
-        credentials = service_account.Credentials.from_service_account_file(
-            str(self.service_account_file),
-            scopes=AUTH_SCOPES,
-        )
-        request = google.auth.transport.requests.Request()
-        credentials.refresh(request)
-        return credentials.token
+        if self.api_version is APIVersion.Legacy:
+            return "key=%s" % (self.api_key,)
+        else:
+            assert self.credentials is not None
+            if not self.credentials.valid:
+                await Deferred.fromFuture(
+                    asyncio.ensure_future(
+                        self.credentials.refresh(self.google_auth_request)
+                    )
+                )
+            return "Bearer %s" % self.credentials.token
```

Comment:
A possible fix for the failing/hanging test, would be to change this as follows:
```suggestion
            await self._refresh_credentials()
            return "Bearer %s" % self.credentials.token
            
    async def _refresh_credentials(self) -> None:
        assert self.credentials is not None
        if not self.credentials.valid:
            await Deferred.fromFuture(
                asyncio.ensure_future(
                    self.credentials.refresh(self.google_auth_request)
                )
            )
```

Then add to `TestGcmPushkin`:
```python
async def _refresh_credentials(self) -> None:
    assert self.credentials is not None
    await self.credentials.refresh(None)
```

... and change the new test assertion from `gcm.last_request_headers` to `notification_req[3]`, since `_request_dispatch` is mocked out in that test case, so `last_request_headers` will never be set.

Reply:
Yes good call I'll do something like that, I guess there is a proper fix but since we can't find it this will do the trick !